File: app/src/main/java/net/tlalka/android/fiszki/models/dao/WordDao.java
package net.tlalka.android.fiszki.models.dao;

import com.j256.ormlite.support.ConnectionSource;
import net.tlalka.android.fiszki.models.entities.Cluster;
import net.tlalka.android.fiszki.models.entities.Lesson;
import net.tlalka.android.fiszki.models.entities.Word;
import net.tlalka.android.fiszki.models.types.LanguageType;

import java.sql.SQLException;
import java.util.List;

public class WordDao extends AbstractDao<Word, Long> {

    public WordDao(ConnectionSource connectionSource) throws SQLException {
        super(connectionSource, Word.class);
    }

    public List<Word> getWordsBy(Lesson lesson) throws SQLException {
        return super.queryBuilder()
                .distinct()
                .where()
                .eq("lesson_id", lesson)
                .query();
    }

    public Word getWordBy(Cluster cluster, LanguageType languageType) throws SQLException {
        return super.queryBuilder()
                .where()
                .eq("cluster_id", cluster)
                .and()
                .eq("languageType", languageType)
                .queryForFirst();
    }
}

File: app/src/main/kotlin/net/tlalka/fiszki/model/dao/WordDao.kt
package net.tlalka.fiszki.model.dao

import com.j256.ormlite.support.ConnectionSource
import net.tlalka.fiszki.model.entities.Cluster
import net.tlalka.fiszki.model.entities.Lesson
import net.tlalka.fiszki.model.entities.Word
import net.tlalka.fiszki.model.types.LanguageType

class WordDao constructor(connectionSource: ConnectionSource)
    : AbstractDao<Word, Long>(connectionSource, Word::class.java) {

    fun getWordBy(cluster: Cluster, languageType: LanguageType): Word {
        return super.queryBuilder()
                .where()
                .eq("cluster_id", cluster)
                .and()
                .eq("languageType", languageType)
                .queryForFirst()
    }

    fun getWordsBy(lesson: Lesson): List<Word> {
        return super.queryBuilder()
                .distinct()
                .where()
                .eq("lesson_id", lesson)
                .query()
    }

    fun getWordsBy(cluster: Cluster): List<Word> {
        return super.queryBuilder()
                .distinct()
                .where()
                .eq("cluster_id", cluster)
                .query()
    }

    fun getWordsBy(languageType: LanguageType): List<Word> {
        return super.queryBuilder()
                .orderBy("value", true)
                .where()
                .eq("languageType", languageType)
                .query()
    }
}
